gapis/trace/android/adreno/validate.go

```diff
@@ -0,0 +1,77 @@
+// Copyright (C) 2019 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package adreno
+
+import (
+	"context"
+
+	"github.com/google/gapid/gapis/perfetto"
+	perfetto_service "github.com/google/gapid/gapis/perfetto/service"
+	"github.com/google/gapid/gapis/trace/android/validate"
+)
+
+var (
+	counters = []validate.GpuCounter{
+		{1, "Clocks / Second", and(isNumber, checkLargerThanZero)},
+		{4, "GPU % Bus Busy", and(isNumber, checkLargerThanZero)},
+		{20, "% Shaders Busy", and(isNumber, checkLargerThanZero)},
+	}
+)
+
+type AdrenoValidator struct {
+}
+
+type checker func(column *perfetto_service.QueryResult_ColumnValues, columnType perfetto_service.QueryResult_ColumnDesc_Type) bool
+
+func and(c1, c2 checker) checker {
+	return func(column *perfetto_service.QueryResult_ColumnValues, columnType perfetto_service.QueryResult_ColumnDesc_Type) bool {
+		return c1(column, columnType) && c2(column, columnType)
+	}
+}
+
+func isNumber(column *perfetto_service.QueryResult_ColumnValues, columnType perfetto_service.QueryResult_ColumnDesc_Type) bool {
+	if columnType != perfetto_service.QueryResult_ColumnDesc_LONG && columnType != perfetto_service.QueryResult_ColumnDesc_DOUBLE {
+		return false
+	}
+	return true
+}
+
+func checkLargerThanZero(column *perfetto_service.QueryResult_ColumnValues, columnType perfetto_service.QueryResult_ColumnDesc_Type) bool {
+	longValues := column.GetLongValues()
+	doubleValues := column.GetDoubleValues()
+	for i := 0; i < validate.SampleCounter; i++ {
+		if columnType == perfetto_service.QueryResult_ColumnDesc_LONG {
+			if longValues[i] <= 0 {
+				return false
+			}
+		} else if columnType == perfetto_service.QueryResult_ColumnDesc_DOUBLE {
+			if doubleValues[i] <= 0.0 {
+				return false
+			}
+		}
+	}
+	return true
+}
```

Comment:
I would
1) make the checker and your three implementations public
2) move all this into the common package.